### ISSUE TITLE
Poor performance when combining many blockwise operations

### ISSUE THREAD

**Issue by ian-r-rose** [dask/dask#9085]
## The Problem

I was doing some performance tests and ran into some pretty poor behavior when a lot (>100) of `Blockwise` operations are applied in series. The same example framed with a single `map_partitions` call performs *much* better. A minimal reproducer:

```python
import distributed
import dask.dataframe as dd
from dask.datasets import timeseries

# A dataframe with a lot of columns to mess with
ddf = timeseries(
    dtypes={
        **{f"name-{i}": str for i in range(25)},
        **{f"price-{i}": float for i in range(25)},
        **{f"id-{i}": int for i in range(25)},
        **{f"cat-{i}": "category" for i in range(25)},
    },
    freq="5s"
)

def transform(df):
    """
    Transform the columns of a pandas or dask dataframe.
    """
    transform = {}
    for c in df.columns:
        dtype = str(df[c].dtype)
        if dtype == "object":
            transform[c] = df[c].str.upper()
        elif dtype == "int64":
            transform[c] = df[c] * 2
        elif dtype == "float64":
            transform[c] = df[c] / 2
    return df.assign(**transform)

ddf.map_partitions(transform).compute()  # does okay!
transform(ddf).compute()  # does *very* badly

```

This reproduces on both local and distributed schedulers, but a comparison of the two from the dask dashboard is shown here (purple is `map_partitions`, teal uses the top-level dataframe API):

![image](https://user-images.githubusercontent.com/5728311/168383651-0d08100b-e479-44b4-a6d6-818b9939b9b5.png)

Not only does using the top-level Dask APIs take several times longer, but it also takes several times as much memory, and it's not hard to knock over workers with the second approach.

A bit of exploration and reflection reveals what's going on: the `Blockwise` layers are being correctly fused into a single layer as we want during optimization. But the resulting local subgraph is pretty complex, and has a ton of intermediate results! If I drill into the layer and plot the subgraph, I get the following:
![subgraph](https://user-images.githubusercontent.com/5728311/168386197-cd839a4c-3aa7-41e0-bfe6-d33f1f0d359b.png)

The performance characteristics of this graph are way worse than just bundling them into the one `map_partitions` call and letting pandas handle intermediate results internally.

Does this matter? I think so. I don't think what I'm doing is too far from "best practices". It is embarrassingly parallel, has immutable operations, and batches all of my transformations together into a single `compute()` call. From an API design perspective, I want to be able to tell users to use the top level API and use `map_partitions` as an escape hatch. I'm not really sure what to do about this behavior right now. It's not necessarily a bug, since `Blockwise` is more-or-less doing the right thing. But it's performing so much worse that I thought I'd write this up for discussion.

### Appendix: why am I using `DataFrame.assign?

I don't see too many people using `DataFrame.assign` like I am here. But I like it for a few reasons:
* It is immutable as opposed to `__setitem__`
* It makes it easier to use method chaining
* I understand it, and `__setitem__` is too complicated.

**Comment by rjzamora** [comment]
>I want to be able to tell users to use the top level API and use map_partitions as an escape hatch

I personally view `map_partitions` as both and "escape hatch" and an optimization. I agree that we should treat pathological cases like this seriously, but I have a feeling we may have better luck encouraging the use of `map_partitions` when the user/dev **knows** munual fusion makes the most sense .

**Comment by ian-r-rose** [comment]
> but I have a feeling we may have better luck encouraging the use of map_partitions when the user/dev knows munual fusion makes the most sense .

Yeah, that's fair. I have a general desire to make pandas best-practices also work with dask dataframes, but it may be unrealistic to expect too much. That being said, I don't think computations like these (lots of embarrassingly parallel transformations) are particularly unusual in an ETL context, so it's a pathological case that I expect to see in the wild.

**Comment by gjoseph92** [comment]
I definitely think of this as a bug. Would applying standard graph optimizations to the SubgraphCallable help here? Or is the subgraph being constructed in some pathological way by Blockwise fusion?
